src/engine/SmtCore.cpp

```diff
@@ -397,6 +397,58 @@ PiecewiseLinearConstraint *SmtCore::chooseViolatedConstraintForFixing( List<Piec
     return candidate;
 }
 
+void SmtCore::replayStackEntry( StackEntry * stackEntry )
+{
+    struct timespec start = TimeUtils::sampleMicro();
+
+    if ( _statistics )
+    {
+        _statistics->incNumSplits();
+        _statistics->incNumVisitedTreeStates();
+    }
+
+    // Obtain the current state of the engine
+    EngineState *stateBeforeSplits = new EngineState;
+    stateBeforeSplits->_stateId = _stateId;
+    ++_stateId;
+    _engine->storeState( *stateBeforeSplits, true );
+    stackEntry->_engineState = stateBeforeSplits;
+
+    // Apply all the splits
+    _engine->applySplit( stackEntry->_activeSplit );
+    for ( const auto &impliedSplit : stackEntry->_impliedValidSplits )
+        _engine->applySplit( impliedSplit );
+
+    _stack.append( stackEntry );
+
+    if ( _statistics )
+    {
+        _statistics->setCurrentStackDepth( getStackDepth() );
+        struct timespec end = TimeUtils::sampleMicro();
+        _statistics->addTimeSmtCore( TimeUtils::timePassed( start, end ) );
+    }
+}
+
+void SmtCore::storeSmtState( SmtState & smtState )
+{
+    smtState._impliedValidSplitsAtRoot = _impliedValidSplitsAtRoot;
+
+    for ( auto &stackEntry : _stack )
+        smtState._stack.append( duplicateStackEntry( *stackEntry ) );
+}
+
+StackEntry * SmtCore::duplicateStackEntry( const StackEntry & stackEntry )
+{
+    StackEntry *copy = new StackEntry();
+
+    copy->_activeSplit = stackEntry._activeSplit;
+    copy->_impliedValidSplits = stackEntry._impliedValidSplits;
+    copy->_alternativeSplits = stackEntry._alternativeSplits;
+    copy->_engineState = NULL;
```

Comment:
Why is the engine state not duplicated?

Reply:
Discussed offline. Added some clarification.